rewrite-java/src/main/java/org/openrewrite/java/tree/J.java

```diff
@@ -3048,6 +3048,79 @@ public InstanceOf withExpr(JRightPadded<Expression> expression) {
         }
     }
 
+    @FieldDefaults(makeFinal = true, level = AccessLevel.PRIVATE)
+    @EqualsAndHashCode(callSuper = false, onlyExplicitlyIncluded = true)
+    @RequiredArgsConstructor
+    @AllArgsConstructor(access = AccessLevel.PRIVATE)
+    final class DeconstructionPattern implements J, Pattern, TypedTree {
+
+        @Nullable
+        @NonFinal
+        transient WeakReference<Padding> padding;
+
+        @With
+        @EqualsAndHashCode.Include
+        @Getter
+        UUID id;
+
+        @With
+        @Getter
+        Space prefix;
+
+        @With
+        @Getter
+        Markers markers;
+
+        @With
+        @Getter
+        JRightPadded<Expression> deconstructor;
+
+        @With
+        @Getter
+        JContainer<Pattern> nested;
```

Comment:
Both of these properties need custom getters and withers, which expose the unwrapped types. I.e. `Expression` for `deconstructor` and `List<Pattern>` for `nested.

Reply:
Actually,. I think there is no need for `deconstructor` to use `JRightPadded`. The space after the type name will be part of the `JContainer`s `before`.

Reply:
Hmm. I suspect this only works correctly due to Java's type erasure. In a pattern like `Point(int x, int y)` the `int x` and `int y` part are mapped to `J.VariableDeclarations` which doesn't implement the `Pattern` interface.

Reply:
We run into the same issue here as with switch pattern matching, should it be a `JContainer<J>` then?

Reply:
Unsure. Having J.VariableDeclarations implement Pattern might be OK. It might cause issues with the inherited getCoordinates() method.

Reply:
Yeah it does cause that issue, so I'm leaning towards changing it to J

Reply:
Unless we want to introduce a `J.BindingPattern` which implements pattern and merely holds a `VariableDeclaration`